Re-enable Kubelet Pod Readiness Probes on Termination and Pod probes should be handled by pod worker

rphillips opened the pull request:
#### What type of PR is this?
/kind bug
/kind regression

#### What this PR does / why we need it:
This PR builds off of @smarterclayton 's #107893 PR to fix `kubelet: Pod probes should be handled by pod worker`. The pod refactor in 2021 disabled [all probes on pod deletion](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2272-L2275). @deads2k also reported to me an [issue](https://github.com/kubernetes/kubernetes/pull/110115) where Pods were failing to report readiness status on termination. Readiness probes are important for controllers and load balancers to know when to take a pod out of service. Not reporting readiness can cause a load balancer (or controller) to react slower to a terminating pod producing 500s and other ill effects. This PR now reports the correct readiness status during termination. 

#107893 needed a test case, which this PR adds. 

#### Which issue(s) this PR fixes:
Fixes #110309

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Fix 1.22 regression so pods will now post their readiness during termination.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

smarterclayton committed:
kubelet: Pod probes should be handled by pod worker

The pod worker is the owner of when a container is running or not,
and the start and stop of the probes for a given pod should be
handled during the pod sync loop. This ensures that probes do not
continue running even after eviction.

Because the pod semantics allow lifecycle probes to shorten grace
period, the probe is removed after the containers in a pod are
terminated successfully. As an optimization, if the pod will have
a very short grace period (0 or 1 seconds) we stop the probes
immediately to reduce resource usage during eviction slightly.

After this change, the probe manager is only called by the pod
worker or by the reconcile loop.

rphillips committed:
kubelet: only shutdown probes for pods that are terminated

This fixes a bug where terminating pods would not run their readiness
probes. Terminating pods are found within the possiblyRunningPods map.

rphillips committed:
kubelet: add e2e test to verify probe readiness

harche commented:
/test pull-kubernetes-node-crio-e2e

rphillips commented:
cc @aojea

matthyx commented:
/triage accepted

aojea commented on test/e2e_node/node_probe.go:
```diff
@@ -0,0 +1,88 @@
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Image:   imageutils.GetE2EImage(imageutils.Perl),
```
can we use Agnhost image instead?

rphillips replied:
Fixed

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -1766,10 +1772,16 @@ func (kl *Kubelet) syncTerminatingPod(ctx context.Context, pod *v1.Pod, podStatu
 	kl.statusManager.SetPodStatus(pod, apiPodStatus)
 
 	if gracePeriod != nil {
+		if *gracePeriod <= 1 {
+			// If we plan to terminate quickly, stop probes immediately, otherwise we will  wait until the pod is completely done
+			kl.probeManager.RemovePod(pod)
```
should we try to optimize this?

rphillips replied:
If they are short grace period pods, then I think this is ok.

smarterclayton replied:
Technically this is to allow a user visible difference between “stop as fast as possible”, and every other kind of stop.  We do that by clamping the grace period to this “less than or equal to one second” range, and since probe periods are quantized by seconds, 1 or below is effectively the same.  And, when someone wants to shut a bunch of pods down as fast as possible, that usually involves freeing up resources or other urgent behavior.

This should also reduce the frequency of noise on shutdown due to probes racing with termination, so it has a net benefit.

I think as long as the comment includes a bit more context about why this is ok, it should be fine.

rphillips replied:
I updated the comment in the latest push. Thanks!

aojea replied:
@smarterclayton I've found the related issue https://github.com/kubernetes/kubernetes/pull/102025 , is 0 still having a special meaning? and we keeping its behavior with this change?

aojea replied:
main concern is this last comment https://github.com/kubernetes/kubernetes/pull/102025#issuecomment-896327232

aojea replied:
I've tried to follow all the graceperiods paths but I'm not able to fully comprehend it , there is this function that seems to always set it to 1
https://github.com/kubernetes/kubernetes/blob/3eadd1a9ead7a009a9abfbd603a5efd0560473cc/pkg/kubelet/pod_workers.go#L641-L671

should we make the optimization at least in a different commit, so we don't mix it with the bugfix?

bobbypage replied:
+1. As @aojea mentioned, perhaps we should include this optimization in a followup PR (or separate commit) entirely. 

https://github.com/kubernetes/kubernetes/pull/102025 did not actually merge so we don't consider "1" as a special grace period and since we plan to backport this PR to 1.23 and 1.22, it probably makes sense to avoid adding this optimization since #102025 was not merged. What do you think @rphillips ? Or is there some unintended side affects without this optimization?

smarterclayton commented on test/e2e_node/node_probe.go:
```diff
@@ -0,0 +1,88 @@
+						{
+							Image:   imageutils.GetE2EImage(imageutils.Agnhost),
+							Name:    podName,
+							Command: []string{"/bin/bash", "-c", "_term() { rm -f /tmp/ready ; sleep 30; exit 0; } ; trap _term SIGTERM; touch /tmp/ready ; while true; do echo \"hello\"; sleep 10; done"},
```
since this is long enough, it would be better to put it in heredoc form

rphillips replied:
Fixed.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -1778,6 +1790,12 @@ func (kl *Kubelet) syncTerminatingPod(ctx context.Context, pod *v1.Pod, podStatu
+	// TODO: once a pod is terminal, certain probes (liveness exec) could be stopped immediately after
+	//   the detection of a container shutdown or (for readiness) after the first failure. Tracked as
+	//   https://github.com/kubernetes/kubernetes/issues/107894 although may not be worth optimizing.
+	kl.probeManager.RemovePod(pod)
```
Adding this invariant to the unit tests would be good.  SyncTerminatingPod should clear probe status, and if kill container returns an error, we should have only stopped liveness probes.

rphillips replied:
Thank you for the feedback. This is a good ide and I added a second test to this PR.

bowei commented:
/cc @bowei

aojea commented:
ok, the bug is worse than I thought, the pods never get unready breaking all apps that depend on the graceful shutdown 

I've verified that the test passes with this fix and fail without.

@rphillips I think that you should incorporate this commit with the e2e tests for Services in this PR
https://github.com/kubernetes/kubernetes/pull/110247/commits/cc6e7857d1f00b9dc5a760fcefcca1420d0b35e3

cc: @deads2k

bobbypage commented:
Hi Ryan! Thanks for your work on this!

I tested the PR locally with the example provided in https://github.com/kubernetes/kubernetes/issues/102537

I don't think it addresses the case mentioned in https://github.com/kubernetes/kubernetes/issues/102537#issue-909658949 (i.e. a pod without a probes).

249043822 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1713,6 +1713,9 @@ func (kl *Kubelet) syncPod(ctx context.Context, updateType kubetypes.SyncPodType
 	pullSecrets := kl.getPullSecretsForPod(pod)
 
+	// Ensure the pod is being probed
+	kl.probeManager.AddPod(pod)
```
https://github.com/kubernetes/kubernetes/blob/c6970e64528ba78b74bf77b86f9b78b7b61bd0cd/pkg/kubelet/prober/prober_manager.go#L188
the `syncPod ` is reentrant, so shall we set the log level above more friendly?

rphillips replied:
This patch does not change the logic within the AddPod function.

rphillips replied:
Good point. I bumped the log level to 4 for these.

rphillips commented:
@bobbypage I looked into pods without readiness probes. From what I can find within the code, K8S has always assumed that a pod without a readiness probe stays in ready=true until it has fully terminated. This PR fixes the case where a pod is defined with a readiness probe, is being terminated, and needs to run the readiness probe on termination.

If we want to change the behavior of readiness for pods without readiness probes, then I believe we will need a different enhancement.

rphillips commented:
Updated the issue to be https://github.com/kubernetes/kubernetes/issues/110309

rphillips committed:
add service e2e tests

bobbypage commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1105,8 +1105,8 @@ func (kl *Kubelet) HandlePodCleanups() error {
-	klog.V(3).InfoS("Clean up probes for terminating and terminated pods")
-	kl.probeManager.CleanupPods(runningPods)
+	klog.V(3).InfoS("Clean up probes for terminated pods")
+	kl.probeManager.CleanupPods(possiblyRunningPods)
```
just for my own understanding - `possiblyRunningPods` refers to pods which are in progress of terminating right? Why do we want need to stop probing here in addition to logic in `syncPod`?

rphillips replied:
HandlePodCleanups() is a goroutine which is running periodically to make sure pods are cleaned up. In my testing, 'runningPods' does not include terminating pods. This would cause the probes to be terminated prematurely (due to the natural race condition between the termination pod_worker logic and HandlePodCleanups).

I didn't want to change the logic of removing the probeManager call from HandlePodCleanups due to unknown side effects.

bobbypage commented:
> @bobbypage I looked into pods without readiness probes. From what I can find within the code, K8S has always assumed that a pod without a readiness probe stays in ready=true until it has fully terminated. This PR fixes the case where a pod is defined with a readiness probe, is being terminated, and needs to run the readiness probe on termination.
> 
> If we want to change the behavior of readiness for pods without readiness probes, then I believe we will need a different enhancement.

ack, makes sense. Thanks for creating dedicated issue to track the particular fix in this PR. small nit: the issue in the pr description links to the PR, not the github issue (#110309) which is what you intend :)

249043822 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1743,6 +1746,9 @@ func (kl *Kubelet) syncTerminatingPod(ctx context.Context, pod *v1.Pod, podStatu
 	// not get invoked again
 	if runningPod != nil {
+		// Stop probing runtime pods immediately
+		kl.probeManager.RemovePod(pod)
```
I also think this line may be useless, the reason is:
1. a runningPod which is synced from runtime could not exists in podmanager, it has no chance to run `syncPod`, so it should has no probes running
2. a runningPod terminating would not update status manager, so it would not report probe results

rphillips replied:
I suspect both are of you a right. I removed this line from the current patch.

rphillips commented on test/e2e_node/node_probe.go:
```diff
@@ -0,0 +1,194 @@
+			items, _ := f.ClientSet.EventsV1().Events(f.Namespace.Name).List(context.Background(), metav1.ListOptions{})
+			for _, event := range items.Items {
+				if strings.Contains(event.Note, "failed liveness probe") {
+					framework.Fail("should not see liveness probe failures")
```
Fixed

SergeyKanzhelev commented:
/assign @bobbypage 
/triage accepted

rphillips commented:
@bobbypage @249043822 I added all the requested changes. Thanks!

bobbypage commented on test/e2e/common/node/container_probe.go:
```diff
@@ -551,6 +553,168 @@ var _ = SIGDescribe("Probing container", func() {
+		framework.ExpectNoError(err)
+		for _, event := range items.Items {
+			// Search only for the pod we are interested in
+			if event.Name != podName {
```
Is this accurate? I thought to filter for events for pod, it was necessary to check involvedObject.name == podName?

bobbypage commented:
One more small question about the test added, thanks for your work on this. 

/lgtm

rphillips commented:
/retest

rphillips commented:
/retest

aojea commented:
/kind regression

folks, this is a regression and has to be backported, pod readiness probes were working in 1.21

https://github.com/kubernetes/kubernetes/pull/110366#issuecomment-1145745415

I still think that the optimization should be a different commit to avoid possible regressions on the backport
https://github.com/kubernetes/kubernetes/pull/110191#discussion_r884282561

rphillips commented:
@dashpole Could you do a final review pass for a lgtm? I'll create backports once we get approvals on this PR.

rphillips commented:
Sorry, @bobbypage ... :)

bobbypage commented on test/e2e/common/node/container_probe.go:
```diff
@@ -551,6 +553,168 @@ var _ = SIGDescribe("Probing container", func() {
+		})
+
+		// verify pods are running and ready
+		err := e2epod.WaitForPodsRunningReady(f.ClientSet, f.Namespace.Name, 1, 0, 90*time.Second, map[string]string{})
```
suggest to use framework's timeout constants f.Timeouts.PodStart instead of custom 90 seconds - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/timeouts.go#L44

bobbypage commented on test/e2e/common/node/container_probe.go:
```diff
@@ -551,6 +553,168 @@ var _ = SIGDescribe("Probing container", func() {
+		podClient.Delete(context.Background(), podName, metav1.DeleteOptions{})
+
+		// Wait for pod to go unready
+		err = wait.PollImmediate(2*time.Second, 30*time.Second, func() (bool, error) {
```
Maybe use `f.Timeouts.PodDelete` here? - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/timeouts.go#L26

bobbypage replied:
replace `2*time.Second` with `f.Poll` - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L113

bobbypage commented on test/e2e/common/node/container_probe.go:
```diff
@@ -551,6 +553,168 @@ var _ = SIGDescribe("Probing container", func() {
+
+		// Shutdown pod. Readiness should change to false
+		podClient.Delete(context.Background(), podName, metav1.DeleteOptions{})
+		err = wait.PollImmediate(2*time.Second, 60*time.Second, func() (bool, error) {
```
replace `2*time.Second` with `f.Poll` - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L113

bobbypage commented on test/e2e/common/node/container_probe.go:
```diff
@@ -551,6 +553,168 @@ var _ = SIGDescribe("Probing container", func() {
+
+		// Verify there are zero liveness failures since they are turned off
+		// during pod termination
+		items, err := f.ClientSet.CoreV1().Events(f.Namespace.Name).List(context.Background(), metav1.ListOptions{})
```
nit: I think there is some latency for events to be generated, so maybe it makes sense to wrap the event check in a `gomega.Consistently` and run it for a bit of time (minute?) to ensure the event won't be generated.

bobbypage commented:
Couple small nits regarding tests and also +1 to the question that @aojea raised here -  https://github.com/kubernetes/kubernetes/pull/110191#discussion_r890569998

rphillips commented:
@bobbypage Thank you for the review.

I updated the tests to the constants you references and added the gomega.Consistently tweak.

I removed the graceperiod optimization since I do not know of any limitations to it.

bobbypage commented:
Thanks for all the fixes and updates!

/lgtm

bobbypage commented:
/assign @mrunalp

rphillips commented:
/assign @aojea and @SergeyKanzhelev

aojea commented:
/approve

Thank you very much
